flowmachine/flowmachine/features/subscriber/entropy.py

```diff
@@ -0,0 +1,391 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Calculates various entropy metrics for subscribers with a specified time
+period.
+"""
+
+from abc import ABCMeta
+from .metaclasses import SubscriberFeature
+from .contact_balance import ContactBalance
+from ..utilities.sets import EventsTablesUnion
+from ..utilities.subscriber_locations import subscriber_locations
+from ...utils.utils import get_columns_for_level
+from ...core import Table
+
+
+class BaseEntropy(SubscriberFeature, metaclass=ABCMeta):
+    """ Base query for calculating entropy of subscriber features. """
+
+    def _make_query(self):
+
+        return f"""
+        SELECT
+            subscriber,
+            -1 * SUM( relative_freq * LN( relative_freq ) ) AS entropy
+        FROM ({self._relative_freq_query}) u
+        GROUP BY subscriber
+        """
+
+    @property
+    def _absolute_freq_query(self):
+
+        raise NotImplementedError
+
+    @property
+    def _relative_freq_query(self):
+        return f"""
+        SELECT
+            subscriber,
+            absolute_freq::float / ( SUM( absolute_freq ) OVER ( PARTITION BY subscriber ) ) AS relative_freq
+        FROM ({self._absolute_freq_query}) u
+        """
+
+
+class PeriodicEntropy(BaseEntropy):
+    """
+    Calculates the recurrence period entropy for events, that is the entropy
+    associated with the period in which events take place. For instance, if
+    events regularly occur at a certain time of day, say at 9:00 and 18:00 then
+    this user will have a low period entropy.
```

Comment:
Can we please include a definition/equation in the docstring, or alternatively a reference to a definition (or both)? Entropy is such an overloaded concept with many slightly varying definitions in different contexts that it will be good for the user to know how we define it here.